Packages/com.ramtype0.meshia.mesh-simplification/Runtime/ProgressiveMeshSimplifyData.cs

```diff
@@ -791,26 +798,52 @@ public void ToMeshData(Mesh.MeshData sourceMesh, Mesh.MeshData destinationMesh,
-
                     }
 
-                    destinationSubMesh.vertexCount = destinationVertexIndex - destinationSubMesh.firstVertex;
                     destinationSubMesh.indexCount = destinationIndexBufferIndex - destinationSubMesh.indexStart;
 
-                    if (destinationSubMesh.vertexCount >= ushort.MaxValue)
-                    {
-                        destinationIndexFormat = IndexFormat.UInt32;
-                    }
-
                     // baseVertex is not set yet
                     destinationSubMeshes[subMeshIndex] = destinationSubMesh;
                 }
             }
 
-
-            for (int destinationVertexIndex = 0; destinationVertexIndex < destinationToSourceVertexIndex.Length; destinationVertexIndex++)
+            for (int sourceVertexIndex = 0, destinationVertexIndex = 0; sourceVertexIndex < sourceVertexCount; sourceVertexIndex++)
             {
-                sourceToDestinationVertexIndex[destinationToSourceVertexIndex[destinationVertexIndex]] = destinationVertexIndex;
+                if (DiscardedVertex.IsSet(sourceVertexIndex))
+                {
+                    sourceToDestinationVertexIndex[sourceVertexIndex] = 0;
+                    continue;
+                }
+                destinationToSourceVertexIndex[destinationVertexIndex] = sourceVertexIndex;
+                sourceToDestinationVertexIndex[sourceVertexIndex] = destinationVertexIndex;
+                uint index = VertexSubMeshIndexes[sourceVertexIndex];
+                for (int subMeshIndex = 0; subMeshIndex < sourceMesh.subMeshCount; subMeshIndex++)
+                {
+                    if ((index & (1u << subMeshIndex)) == 0)
+                    {
+                        continue;
+                    }
+                    ref var destinationSubMesh = ref destinationSubMeshes.ElementAt(subMeshIndex);
+                    if (destinationSubMesh.vertexCount == 0)
+                    {
+                        destinationSubMesh.firstVertex = destinationVertexIndex;
+                        destinationSubMesh.vertexCount = 1;
+                    }
+                    else if (destinationVertexIndex < destinationSubMesh.firstVertex)
+                    {
+                        destinationSubMesh.vertexCount += destinationSubMesh.firstVertex - destinationVertexIndex;
+                        destinationSubMesh.firstVertex = destinationVertexIndex;
+                    }
+                    else if (destinationVertexIndex >= destinationSubMesh.firstVertex + destinationSubMesh.vertexCount)
+                    {
+                        destinationSubMesh.vertexCount = destinationVertexIndex - destinationSubMesh.firstVertex + 1;
+                    }
+                }
+                destinationVertexIndex++;
+                if (destinationVertexIndex >= ushort.MaxValue)
+                {
+                    destinationIndexFormat = IndexFormat.UInt32;
+                }
```

Comment:
Even if we have more than `ushort.MaxValue` vertices in the mesh, we could still use `UInt16` index format if all of sub-meshes has less than `ushort.MaxValue` vertex index range.

Reply:
This checking is from line 800-804 of the original version, I assume the original logic is correct, I moved it here.

Reply:
https://github.com/RamType0/Meshia.MeshSimplification/blob/5dc2dd52bbf6ea192c81f993461cf71508296963/Packages/com.ramtype0.meshia.mesh-simplification/Runtime/ProgressiveMeshSimplifyData.cs#L800-L803

It uses different comparand.